### PR TITLE
[ALLUXIO-3217] Improve Inode Locking

### PR DESCRIPTION
**Pull request by yuzhu**
https://alluxio.atlassian.net/browse/ALLUXIO-3217

The main goal of this PR is to use composable locklist to implement locking all in `LockedInodePath` and eliminate the need for `TempInodePathForChild` and `TempInodePathForDescendant`.

### CODE REVIEW

**Comment by yuzhu** [comment]
jenkins, test this please

**Comment by yuzhu** [comment]
jenkins, test this please

**Comment by yuzhu** [comment]
jenkins, test this please

**Comment by yuzhu** [comment]
jenkins, test this please

**Comment by yuzhu** [comment]
jenkins, test this please

**Comment by yuzhu** [comment]
jenkins, test this please

**Comment by yuzhu** [comment]
jenkins, test this please

**Comment by yuzhu** [comment]
jenkins, test this please

**Comment by yuzhu** [comment]
@gpang  PTAL, I will add more documentation and tests, but would like some high-level comments.

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java [hunk 1]
```diff
@@ -885,19 +905,11 @@ public void setPinned(LockedInodePath inodePath, boolean pinned, long opTimeMs)
     } else {
       assert inode instanceof InodeDirectory;
       // inode is a directory. Set the pinned state for all children.
-      try (TempInodePathForDescendant tempInodePath = new TempInodePathForDescendant(inodePath)) {
+      try {
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: Is this `try` extraneous?

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java [hunk 2]
```diff
@@ -784,41 +807,38 @@ public InodeLockList lockDescendant(LockedInodePath inodePath, LockMode lockMode
...
-        try {
-          inodeGroup.lockWriteAndCheckParent(child, inodeDirectory);
-        } catch (InvalidPathException e) {
-          // Inode is no longer a child, continue.
-          continue;
-        }
+      LockedInodePath lockedDescendantPath;
+      InodeLockList lockList = new InodeLockList();
+      try {
+        lockList.lockReadAndCheckParent(child, inodeDirectory);
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: how is the `lockMode` used?

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java [hunk 3]
```diff
@@ -784,41 +807,38 @@ public InodeLockList lockDescendant(LockedInodePath inodePath, LockMode lockMode
...
+            inodePath, lockList);
+        inodePathList.add(lockedDescendantPath);
+      } catch (InvalidPathException e) {
+        // Inode is no longer a child, continue.
+        continue;
       }
-      if (child.isDirectory()) {
-        lockDescendantsInternal((InodeDirectory) child, lockMode, inodeGroup);
+
+      if (child.isDirectory() && lockedDescendantPath != null) {
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: when would `lockedDescendantPath` be  `null`?

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java [hunk 4]
```diff
@@ -784,41 +807,38 @@ public InodeLockList lockDescendant(LockedInodePath inodePath, LockMode lockMode
...
-          inodeGroup.lockWriteAndCheckParent(child, inodeDirectory);
-        } catch (InvalidPathException e) {
-          // Inode is no longer a child, continue.
-          continue;
-        }
+      LockedInodePath lockedDescendantPath;
+      InodeLockList lockList = new InodeLockList();
+      try {
+        lockList.lockReadAndCheckParent(child, inodeDirectory);
+        lockedDescendantPath = new MutableLockedInodePath(inodePath.getUri().join(child.getName()),
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: Why doesn't this use something like `LockedInodePath#createTempPathForChild`?

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java [hunk 5]
```diff
@@ -624,7 +624,8 @@ public CreatePathResult createPath(RpcContext rpcContext, LockedInodePath inodeP
       }
 
       createdInodes.add(dir);
-      extensibleInodePath.getInodes().add(dir);
+      // No need to add to InodePath because we already add to its locklist
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: How is it already in the locklist?

File: core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java [hunk 6]
```diff
@@ -331,7 +331,7 @@ void rename(AlluxioURI srcPath, AlluxioURI dstPath, RenameOptions options)
   // into RuntimeException on the client.
   void free(AlluxioURI path, FreeOptions options)
       throws FileDoesNotExistException, InvalidPathException, AccessControlException,
-      UnexpectedAlluxioException, UnavailableException;
+      UnexpectedAlluxioException, IOException;
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java]: Why is this changed?

File: core/server/master/src/main/java/alluxio/master/file/meta/LockedInodePath.java [hunk 7]
```diff
@@ -114,32 +121,31 @@ public synchronized InodeDirectory getParentInodeDirectory()
...
-    return mInodes.get(mInodes.size() - 1);
+    return mLockList.getInodes().get(mLockList.getInodes().size() - 1);
   }
-
   /**
    * @return a copy of the list of existing inodes, from the root
    */
   public synchronized List<Inode<?>> getInodeList() {
-    return Lists.newArrayList(mInodes);
+    return Lists.newArrayList(mLockList.getInodes());
```
[the remark this answers is not in the document]
Comment (yuzhu) [reply core/server/master/src/main/java/alluxio/master/file/meta/LockedInodePath.java]: `CompositeInodeLockList` does but not the base `InodeLockList`, made them both create a copy now.

File: core/server/master/src/main/java/alluxio/master/file/meta/LockedInodePath.java [hunk 8]
```diff
@@ -30,24 +30,30 @@
...
-  protected final ArrayList<Inode<?>> mInodes;
   protected final InodeLockList mLockList;
   protected InodeTree.LockMode mLockMode;
 
-  LockedInodePath(AlluxioURI uri, List<Inode<?>> inodes, InodeLockList lockList,
+  LockedInodePath(AlluxioURI uri, InodeLockList lockList,
       InodeTree.LockMode lockMode)
       throws InvalidPathException {
-    Preconditions.checkArgument(!inodes.isEmpty());
+    Preconditions.checkArgument(!lockList.getInodes().isEmpty());
```
[the remark this answers is not in the document]
Comment (yuzhu) [reply core/server/master/src/main/java/alluxio/master/file/meta/LockedInodePath.java]: made a `locklist#isEmpty` function that is more efficient

File: core/server/master/src/main/java/alluxio/master/file/meta/LockedInodePath.java [hunk 9]
```diff
@@ -30,24 +30,30 @@
...
     mUri = uri;
     mPathComponents = PathUtils.getPathComponents(mUri.getPath());
-    mInodes = new ArrayList<>(inodes);
     mLockList = lockList;
     mLockMode = lockMode;
   }
 
+  LockedInodePath(AlluxioURI descendantUri, LockedInodePath lockedInodePath,
```
[the remark this answers is not in the document]
Comment (yuzhu) [reply core/server/master/src/main/java/alluxio/master/file/meta/LockedInodePath.java]: done. changed the variable name to `locklist` to be more descriptive.

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java (hunk 1 above)
Comment (yuzhu) [reply core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: don't think so, the logic is that if a single `lockDescendantPath` failed, we should not try to  lock more descendants. Hence the `try` block outside of the loop.

File: core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java (hunk 6 above)
Comment (yuzhu) [reply core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java]: Because we now use  `Closer` to close resources, which can throw an `IOException`

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java (hunk 5 above)
Comment (yuzhu) [reply core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: This is not obvious, but line 521 `InodeLockList lockList = traversalResult.getInodeLockList();` actually gets the same locklist that is in the `LockedInodePath` . So when we use the lockList to lock new Inodes, it adds it to the `LockedInodePath`

**Comment by gpang** [comment]
Jenkins, test this please

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 10]
```diff
@@ -1067,17 +1065,26 @@ public FileSystemMasterView getFileSystemMasterView() {
...
       syncMetadata(rpcContext, parent, lockingScheme, DescendantType.ALL);
 
-      try (InodeLockList children = mInodeTree.lockDescendants(parent, InodeTree.LockMode.READ)) {
-        if (!checkConsistencyInternal(parent.getInode(), parent.getUri())) {
-          inconsistentUris.add(parent.getUri());
+      if (!checkConsistencyInternal(parent.getInode(), parent.getUri())) {
+        inconsistentUris.add(parent.getUri());
+      }
+
+      List<LockedInodePath> children = mInodeTree.lockDescendants(parent, InodeTree.LockMode.READ);
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: This pattern is found in several places. Maybe we should have a new class which is closeable, so we can just use try-with-resources on the list.

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java [hunk 11]
```diff
@@ -749,7 +772,7 @@ public long reinitializeFile(LockedInodePath inodePath, long blockSizeBytes, lon
    * @throws FileDoesNotExistException if inode does not exist
    */
   public InodeLockList lockDescendant(LockedInodePath inodePath, LockMode lockMode,
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: Can this be `private`, or just incorporated into `lockDescendantPath`?

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java [hunk 12]
```diff
@@ -784,41 +807,49 @@ public InodeLockList lockDescendant(LockedInodePath inodePath, LockMode lockMode
...
+    }
+    InodeDirectory inodeDirectory = (InodeDirectory) inode;
     for (Inode<?> child : inodeDirectory.getChildren()) {
+      LockedInodePath lockedDescendantPath;
+      InodeLockList lockList = new InodeLockList();
       if (lockMode == LockMode.READ) {
         try {
-          inodeGroup.lockReadAndCheckParent(child, inodeDirectory);
+          lockList.lockReadAndCheckParent(child, inodeDirectory);
+          lockedDescendantPath = new MutableLockedInodePath(
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: Can this functionality be provided by `LockedInodePath`? A `LockedInodePath` currently provides a new `LockedInodePath` for a non-existing child. There can be a similar method for returning a `LockedInodePath` of an existing child?

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java (hunk 5 above)
Comment (yuzhu) [reply core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: i agree using `extensibleInodePath.getLockList()` makes it less confusing. Not sure about `TraversalResult` returning `LockedInodePath`, I think there is a version of `traverseToInode` that does not take `LockedInodePath` as a parameter, it would be odd to return a `LockedInodePath` in that method.

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java [hunk 13]
```diff
@@ -654,12 +655,17 @@ public CreatePathResult createPath(RpcContext rpcContext, LockedInodePath inodeP
           // The final path component already exists and is not persisted, so it should be added
           // to the non-persisted Inodes of traversalResult.
           syncPersistDirectory(rpcContext, (InodeDirectory) lastInode);
+
         } else if (!lastInode.isDirectory() || !(options instanceof CreateDirectoryOptions
             && ((CreateDirectoryOptions) options).isAllowExists())) {
           String errorMessage = ExceptionMessage.FILE_ALREADY_EXISTS.getMessage(path);
           LOG.error(errorMessage);
+          lockList.unlockLast();
```
[the remark this answers is not in the document]
Comment (yuzhu) [reply core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: It is the same reason we have to unlock on line 666 below.

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java (hunk 10 above)
Comment (yuzhu) [reply core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: are you suggesting we make a `LockedInodePathList` class that is closeable?

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java (hunk 11 above)
Comment (yuzhu) [reply core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: made `private`

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java (hunk 12 above)
Comment (yuzhu) [reply core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: done

**Comment by yuzhu** [comment]
jenkins, test this please

**Comment by yuzhu** [comment]
@gpang addressed comments, PTAL.

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java (hunk 10 above)
Comment (gpang) [reply core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: Yeah, maybe? Would that simplify the usage of `lockDescendants`?

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java (hunk 5 above)
Comment (gpang) [reply core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: ack.

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java [hunk 14]
```diff
@@ -518,14 +518,13 @@ public CreatePathResult createPath(RpcContext rpcContext, LockedInodePath inodeP
     LOG.debug("createPath {}", path);
 
     TraversalResult traversalResult = traverseToInode(inodePath, inodePath.getLockMode());
-    InodeLockList lockList = traversalResult.getInodeLockList();
-
+    //InodeLockList lockList = traversalResult.getInodeLockList();
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: you can remove this line.

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java [hunk 15]
```diff
@@ -709,7 +715,7 @@ public CreatePathResult createPath(RpcContext rpcContext, LockedInodePath inodeP
         rpcContext.getJournalContext().append(lastInode.toJournalEntry());
 
         createdInodes.add(lastInode);
-        extensibleInodePath.getInodes().add(lastInode);
+        // extensibleInodePath.getInodes().add(lastInode);
```
[the remark this answers is not in the document]
Comment (gpang) [reply core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: You can remove this line.

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java [hunk 16]
```diff
@@ -654,12 +658,17 @@ public CreatePathResult createPath(RpcContext rpcContext, LockedInodePath inodeP
...
           syncPersistDirectory(rpcContext, (InodeDirectory) lastInode);
+
         } else if (!lastInode.isDirectory() || !(options instanceof CreateDirectoryOptions
             && ((CreateDirectoryOptions) options).isAllowExists())) {
           String errorMessage = ExceptionMessage.FILE_ALREADY_EXISTS.getMessage(path);
           LOG.error(errorMessage);
+          extensibleInodePath.getLockList().unlockLast();
           throw new FileAlreadyExistsException(errorMessage);
         }
+        // We need to remove the last inode from the locklist because it is an inode
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: can you move this up to https://github.com/Alluxio/alluxio/pull/7342/files#diff-3ae46d8ad34ec27f13f732e3155d2d0eR655 so it is done for all code paths when the inode already exists?

Also, can you comment why it was already locked (it was locked during traversal, and locked again here).

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeLockList.java [hunk 17]
```diff
@@ -167,7 +169,14 @@ public synchronized void lockWriteAndCheckNameAndParent(Inode<?> inode, Inode pa
    * @return the list of inodes locked in this lock list, in order of when the inodes were locked
    */
   public synchronized List<Inode<?>> getInodes() {
-    return mInodes;
+    return Lists.newArrayList(mInodes);
```
[the remark this answers is not in the document]
Comment (gpang) [reply core/server/master/src/main/java/alluxio/master/file/meta/InodeLockList.java]: Is it possible to make this API not make a copy, but only let the callers make the copy if necessary?

[the remark this answers is not in the document]
Comment (yuzhu) [reply core/server/master/src/main/java/alluxio/master/file/meta/InodeLockList.java]: Missed this one,  its subclass `CompositeInodeLockList` will have to return a copy, so I figured it was more consistent to return a copy in the parent class as well.

**Comment by yuzhu** [comment]
jenkins, test this please

**Comment by yuzhu** [comment]
@gpang  PTAL, comments addressed.

File: core/server/master/src/main/java/alluxio/master/file/meta/LockedInodePath.java [hunk 18]
```diff
@@ -75,7 +90,8 @@ public synchronized AlluxioURI getUri() {
     if (!fullPathExists()) {
       return null;
     }
-    return mInodes.get(mInodes.size() - 1);
+    List<Inode<?>> inodeList = Lists.newArrayList(mLockList.getInodes());
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/LockedInodePath.java]: why is there a `newArrayList`?

File: core/server/master/src/main/java/alluxio/master/file/meta/LockedInodePath.java [hunk 19]
```diff
@@ -177,14 +192,72 @@ public synchronized void downgradeLastWithScheme(LockingScheme lockingScheme) {
...
+      throws InvalidPathException {
+    Preconditions.checkNotNull(getInodeOrNull());
+    Preconditions.checkState(getInodeOrNull().isDirectory(),
+        "Trying to create TempPathForChild for a file inode");
+    return new MutableLockedInodePath(mUri.join(childName), new CompositeInodeLockList(mLockList),
+        mLockMode);
+  }
+
+  /**
+   * Constructs a temporary {@link LockedInodePath} from an existing {@link LockedInodePath}, for
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/LockedInodePath.java]: The comments should distinguish between `createTempPathForChild` and `createTempPathForExistingChild`. I think this one is for an existing inode, and the other one just adds to the path, so the inode is not locked yet (could not exist).

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java [hunk 20]
```diff
@@ -784,41 +810,35 @@ public InodeLockList lockDescendant(LockedInodePath inodePath, LockMode lockMode
...
-      throws FileDoesNotExistException {
-    Inode<?> inode = inodePath.getInode();
-    InodeLockList inodeGroup = new InodeLockList();
-    if (!inode.isDirectory()) {
-      return inodeGroup;
-    }
-    return lockDescendantsInternal((InodeDirectory) inode, lockMode, inodeGroup);
+  public LockedInodePathList lockDescendants(LockedInodePath inodePath, LockMode lockMode) {
+    List<LockedInodePath> inodePathList = new ArrayList<>();
+    lockDescendantsInternal(inodePath, lockMode, inodePathList);
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: Is there any chance that this throws an exception? Then we could have some orphan `LockedInodePath`s which are never closed.

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java [hunk 21]
```diff
@@ -619,12 +619,13 @@ public CreatePathResult createPath(RpcContext rpcContext, LockedInodePath inodeP
...
           // After creation and journaling, downgrade to a read lock.
-          lockList.downgradeLast();
+          extensibleInodePath.getLockList().downgradeLast();
         }
       }
 
       createdInodes.add(dir);
-      extensibleInodePath.getInodes().add(dir);
+      // No need to add to InodePath because we already add to its locklist
+      // extensibleInodePath.getInodes().add(dir);
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: please remove this comment.

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeLockList.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (gpang) [reply core/server/master/src/main/java/alluxio/master/file/meta/InodeLockList.java]: I see. I just worry about the frequent objects this will create. At the very least, we should add a TODO to make that API not return a copy.

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 22]
```diff
@@ -2400,38 +2407,42 @@ private void freeAndJournal(RpcContext rpcContext, LockedInodePath inodePath, Fr
     long opTimeMs = System.currentTimeMillis();
     List<Inode<?>> freeInodes = new ArrayList<>();
     freeInodes.add(inode);
+    List<LockedInodePath> descendants = mInodeTree.lockDescendants(inodePath,
+        InodeTree.LockMode.WRITE).getInodePathList();
+    Closer closer = Closer.create();
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: why does this method still need the closer?

File: core/server/master/src/main/java/alluxio/master/file/meta/LockedInodePath.java (hunk 18 above)
Comment (yuzhu) [reply core/server/master/src/main/java/alluxio/master/file/meta/LockedInodePath.java]: done

File: core/server/master/src/main/java/alluxio/master/file/meta/LockedInodePath.java (hunk 19 above)
Comment (yuzhu) [reply core/server/master/src/main/java/alluxio/master/file/meta/LockedInodePath.java]: addressed

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java (hunk 20 above)
Comment (yuzhu) [reply core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: Should not throw an exception, since we catch the exceptions in `lockDescendantsInternal` and handle them.

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java (hunk 22 above)
Comment (yuzhu) [reply core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: Because using a closer is more flexible.  In this particular case, we add the inodePath to the list after  we generate the list, so we don't actually want to close every elements in the array.

**Comment by yuzhu** [comment]
@gpang  addressed comments, PTAL. Thanks

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 23]
```diff
@@ -2400,38 +2407,42 @@ private void freeAndJournal(RpcContext rpcContext, LockedInodePath inodePath, Fr
     long opTimeMs = System.currentTimeMillis();
     List<Inode<?>> freeInodes = new ArrayList<>();
     freeInodes.add(inode);
+    List<LockedInodePath> descendants = mInodeTree.lockDescendants(inodePath,
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: Could you add a comment on why this is not using the `lockDescendants` closable resource, so we know why it does not follow the convention.

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeLockList.java [hunk 24]
```diff
@@ -167,7 +169,14 @@ public synchronized void lockWriteAndCheckNameAndParent(Inode<?> inode, Inode pa
    * @return the list of inodes locked in this lock list, in order of when the inodes were locked
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeLockList.java]: This should mention that this is a new/copy list. Also, please add a TODO to change this API to not return a copy.

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java (hunk 23 above)
Comment (yuzhu) [reply core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: done

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeLockList.java (hunk 24 above)
Comment (yuzhu) [reply core/server/master/src/main/java/alluxio/master/file/meta/InodeLockList.java]: done

**Comment by yuzhu** [comment]
@gpang comments addressed, PTAL. Thanks

**Comment by gpang** [comment]
Jenkins, test this please
